Export reqwest crate for easier usage of SlackWebRequestSender impl

bwasty opened the pull request:
Trying to use a 'directly imported' `reqwest::Client` with slack-rs-api might fail if your `reqwest` version doesn't match the one used by slack-rs-client (see also slack-rs/slack-rs#81). Re-exporting should help with that.

mthjones commented:
Thanks for the PR! The use-case makes sense to me, but I'm not sure about re-exporting the entirety of the reqwest crate. I like the approach used in `rusoto` where they export a `default_client()` function that creates the "default", but users still have the flexibility to implement the trait themselves if they have a newer version. I'd also be fine with just re-exporting `reqwest::Client`, though that's harder to document the usage of properly.

bwasty commented:
I had already changed it to only export `reqwest::Client`, but then noticed I needed another part of reqwest in my bot (`reqwest::header::{Authorization, Bearer};`, to download slack files). With the full export I can drop reqwest from my Cargo.toml.

I'll have a look at rusoto.

mthjones commented:
> I needed another part of reqwest in my bot (reqwest::header::{Authorization, Bearer};, to download slack files

Can you expand a bit more on that, perhaps with some sample code of what you're doing? I'm interested if this is a use-case that isn't supported very well in the library or if it's outside the scope of the library.

bwasty commented:
My bot is listening for FileShare messages, downloading the file, processing it and posting a "summary" to another channel. Here's a condensed snippet from my code: 
```rust
        if let Message::FileShare(MessageFileShare { file, text, .. }) = message {
            let file = file.unwrap();
            let response = self.reqwest_client.get(&file.url_private.unwrap())
                .header(Authorization(Bearer { token: self.token.clone() }))
                .send();
            let mut contents = String::new();
            response.read_to_string(&mut contents));
            // ...
        }`
```

Downloading the file works differently than the rest of the API as far as I see, it needs an Authorization header ([docs](https://api.slack.com/tutorials/working-with-files)). Might actually be nice to add this to the library.

mthjones commented:
Ah, I see, so it's not actually going through the API. In that case, I'd suggest that your code having a dependency on `reqwest` is entirely justified.

The default client here (which I agree with it getting exported!) is mostly just to help with the "simple" case where a user just wants to send Slack messages without having to think too much about setting up an HTTP client. If you want to share a client for sending other requests, I'd suggest implementing `SlackWebRequestSender` for your `reqwest` version's Client.

I think I did a poor job of documenting that. I'll open an issue to improve documentation around using clients.

bwasty commented:
> If you want to share a client for sending other requests, I'd suggest implementing SlackWebRequestSender for your reqwest version's Client.

I'm not sure that's possible - I tried that before finding the mismatching version, but I got
```only traits defined in the current crate can be implemented for arbitrary types
impl doesn't use types inside crate

note: the impl does not reference any types defined in this crate
note: define and implement a trait or new type instead
```

mthjones commented:
Argh, yeah, that's a good point. You'd have to wrap the client in a newtype in order to implement it. It's kind of a pain... but it's a problem any library is going to have that wants to let users provide their own clients.

Here's a hypothetical. Say you wanted your bot to now integrate with AWS, so you pull in `rusoto` as a dependency. You'll probably want to use the same request sender with rusoto. If this exported the `reqwest` crate and `rusoto` did the same, which one do you use? Presumably whichever exports the later version, then you'd implement the trait from the other crate for the exported client. However, now you need to wrap the exported client in a newtype in order to implement the other trait, then you end up back at this same issue.

I guess what I'm saying is, there's really only two scenarios that are reasonable to support: the simple "I just want something that works" option (default client) and the "I want to use this client for other things as well" option. Re-exporting the crate only helps with one specific scenario of the latter, where you want to send your own requests + send Slack requests, but falls apart pretty quickly once you bring in other things that also want to send requests generically.

bwasty commented:
I added the `default_client()` function and changed the exports (I pushed a few intermediate versions so I could test with my bot and the slack-rs example in case you wondered). Wasn't sure how to deal with the error type - I think it has to be exported too so it can be matched against.

mthjones commented:
Looks good to me. Thanks!
